### ISSUE TITLE
-:!> in a wet gate feels unintuitive

### ISSUE THREAD

**Issue by techieAgnostic** [urbit/urbit#1347]
When using -:!> in a wet gate, it will return the type specified in the gate, rather than the type passed into the gate at the call end, contrary to how a wet gate appears to handle the type it returns.

not 100% sure if this is an issue or not, but if theres a reason for it to act like this I'd be interested in hearing about it.

Thanks to all the contributors

-tA

**Comment by joshuareagan** [comment]
For some examples to illustrate...

Here is the type spear behavior inside the wet gate:

```
> =wetgate |*(a=@ [-:!>(a) a])

> (wetgate 12)
[#t/@ 12]

> (wetgate 0xbeef)
[#t/@ 0xbeef]
```

Here's how it behaves outside the wet gate:

```
> -:!>((|*(a=@ a) 12))
#t/@ud

> -:!>((|*(a=@ a) 0xbeef))
#t/@ux
```

The latter examples seem fine. But the type spear is using the bunt sample type for the type spear result, and not the call-site sample. This seems weird. In fact, it means that the type spear isn't guaranteed to show what the runtime type is. E.g.:

```
> (|*(a=@ [-:!>(a) a]) [12 14])
[#t/@ 12 14]
```

**Comment by philipcmonk** [comment]
Hmm, that does seem problematic.

The reified type goes straight in the nock (produced in +burp), so if it varied based on the call-site type, it would violate the constraint that wet gates must produce the same nock regardless of call-site type (it would be a sort of reflection and would allow ad-hoc polymorphism).  Or rather, it would just fail that check, so using a `!>` inside a wet gate on something involving the sample would effectively monomorphize the gate so that you could only call it with the formal sample type (but it wouldn't coerce your input to that type like a dry gate, so you'd always need to cast your input to that type).

The current behavior is, I think, worse than that, because of @joshuareagan 's last example, which is Very Bad.

I'd say we should make sure the result of the `!>` is the same at define-site and call-site, and type fail if it's not.

**Comment by joemfb** [comment]
That last example is part of a more general problem: wet gate call seem to be missing an obvious nest check (that the actual sample nests within the bunted sample).

```
> (|*(a=@ a) [%foo %bar])
[%foo %bar]
```

I can't think of any use-case that would require we elide that check, but I also haven't tried adding it.

As @philipcmonk notes, given the invariants of wet gates (ie, they must produce the same nock for all call sites), !> in a wet gate will always have to produce the type of the bunted sample.

**Comment by joshuareagan** [comment]
I haven't studied the wet gate part of the compiler enough to feel too confident in saying this, but my inclination is to agree with @joemfb. It seems to me that a call-site type check of the argument against the sample type would be the simplest and most intuitive fix. If you really want more values for your wet gate, just pick a broader sample type.

**Comment by joshuareagan** [comment]
I hesitate because this solution seems so obvious -- surely there's a reason we haven't added this check before?

**Comment by philipcmonk** [comment]
This wouldn't let you pass wet gates into other wet gates, like `(turn list-of-pairs head)` because wet gates don't nest in anything other than themselves.

**Comment by ohAitch** [comment]
A literal type check would completely break calling wet gates on(anonymous,
gold) gates

In general, the /point/ of wet gates is to call them with types different
than the sample, not just more specific than the sample; my inclination
here is to say that if you're doing anything with vases it it your
responsibility to keep the types in line, especially if you're doing a
convoluted [-:!>(a) +:!>(a)] instead of just !>(a)

That said I do expect just !>(a) to have the same issue, which can be
resolved either by adding a nest check /there/ or forbidding !> in wet
cores entirely

**Comment by philipcmonk** [comment]
`!>` is fine in wet gates if it doesn't involve the sample.  The most consistent answer is to just say "the nock of calling it with the formal sample type must match what you get calling it with the actual sample type", which I thought was the rule anyway.

**Comment by joshuareagan** [comment]
@philipcmonk, unless I'm misunderstanding you I think you're mistaken:

```
> =takeswetgate |*(a=* a)
> (takeswetgate |*(b=* b))
[resulting wet gate shows up here]
```

It's just that `!>` is only going to let you see that gate as a noun. (And yes, it would involve revision of gates like `turn` to make the sample `*`, which is messier than I'd initially thought.)

**Comment by ohAitch** [comment]
That is the rule, but I think there's a !> specific exception due to it
grabbing the entire current subject in a %hold, or something along those
lines

**Comment by philipcmonk** [comment]
@joshuareagan but you can't call the gate, right?  It should give a `find.+2.$` or something

**Comment by joshuareagan** [comment]
You can:

```
> ((takeswetgate |*(b=* b)) [12 14])
[12 14]
```

**Comment by philipcmonk** [comment]
I mean `|*(a=* (a 1))` doesn't compile

**Comment by joshuareagan** [comment]
You're right.

**Comment by philipcmonk** [comment]
Basically, we need enough information to be able to call it, but that's too much information for it to nest.  We don't have a generic "gate" type -- it must specify wet or dry.  That doesn't sound great to me, but I'm not sure what the solution is.

**Comment by joshuareagan** [comment]
idk, something just feels very flimsy and weird about the way we handle wet gate samples -- it seems to me very bad that we purposely allow a totally wrong argument type and then treat it like the sample type inside the wet gate.

It feels like we're missing a feature of the language that would allow us to do a type check against the sample type at the call site and also run the argument as a gate.

**Comment by joshuareagan** [comment]
That is, I think I'm wondering whether wet gates are fundamentally solved yet re: treatment of their samples.

**Comment by joshuareagan** [comment]
For the record I know what a horrible thing I'm saying -- wet gates have been rewritten several times already, and making their argument types behave with the desired sample faces is extremely important and difficult to enforce.

But are we really sure that wet gates have been largely solved?

**Comment by philipcmonk** [comment]
There are several self-consistent treatments of samples.  Dry gates are self-consistent ("if it nests within the formal sample, we assign it the type of the formal sample").  Parametric polymorphism is self-consistent ("it does the same thing regardless of the type of the input" (possibly bounded)), as is ad-hoc polymorphism ("you can check the type of the input and react differently at runtime").  The `|*  a=mold  |=  =a` pattern is isomorphic to unbounded parametric polymorphism.

Wet gates are another treatment of samples, and I think they're self-consistent except for this bug.  They're a form of parametric polymorphism that doesn't rely on subtyping whatsoever with the rule "if the nock of compiling with the formal sample is the same as the nock of compiling it with the actual sample (implying both compile successfully), we're good".  That's a simple enough rule that it's plausibly a good idea.  In practice, the weird face handling breaks the simplicity, and I think they're very problematic, as described in urbit/arvo#1145.  The impedance mismatch between this and dry gates is big enough that they can hardly be called two forms of the same thing.  For example, nest-wise they're not comparable.

IOW, I don't think wet gates have been "solved", but they're very near a local optimum, and I don't think little hacks will solve it -- it needs to switch rules altogether.

**Comment by ohAitch** [comment]
Wet gates exist solely for the purpose of using totally wrong argument types - otherwise you could just go dry!

I would not say they are "solved", just at a good-enough equilibrium of "the footguns are obscure things like this" and "simple enough to follow what they're doing", with `bake` being available if you want more explicit type annotations. In general the topic of handling generics is an area of language design that is being actively researched and in fact has yet to produce a clear winner; I think mostly I'm on board with your "horrible thing" point of trying to do less language design research at urbit, at least for the next decade.

**Comment by ohAitch** [comment]
Are wet gates any less comparable nest-wise than _dry_ gold gates?

**Comment by philipcmonk** [comment]
Yeah, dry gates must match sample, product, and context types, which is pretty restrictive, but wet gates have to match sample, product, and context *values*.

**Comment by philipcmonk** [comment]
Although my point was that no dry gate nests within a wet gate or vice versa, and the only things they both nest within are I think `^` and `*`

**Comment by ohAitch** [comment]
Huh, even if the types are identical(as opposed to just nesting)? I thought that information wasn't like, retained

**Comment by philipcmonk** [comment]
Maybe the context checks on types, I forget.  But the formula definitely has to match by value, precisely because the types of the sample aren't sufficient to guarantee anything about the product (because you can use "totally wrong" argument types)

**Comment by joshuareagan** [comment]
> Wet gates exist solely for the purpose of using totally wrong argument types - otherwise you could just go dry!

This is *not* a good characterization. Wet gates exist to enable type-generic programming (and thus admit generic polymorphism) while leaving the option of preserving the call-site type (or some type related to it). Dry gates are meant to let you specify at the def-site which type you want.

I'm sympathetic to the complaint that we might be talking about a problem that isn't high-priority for people working on Hoon right now -- but I'm not as sympathetic to the claim that these sharp corners are actually features of the language.

**Comment by ohAitch** [comment]
That's fair - I think a more serious version of that claim would be "we do not have a good enough definition of Totally Wrong that enforcing it wouldn't create more issues than it solves". Or rather, we have one, it's dry gates with iron polymorphism, but for the set of problems *that* is insufficient to address(broadly, /user/-level polymorphism).

I definitely would be interested in a version of wet gate sample enforcement that e.g. checks nesting by default but gives up any time it hits a core(besides checking that the other branch is also a core). Assuming that doesn't complain about existing code ~written by people other than me~ that is ;)

**Comment by joshuareagan** [comment]
For the record, here is the code for type-checking cores:

https://github.com/urbit/arvo/blob/master/sys/hoon.hoon#L10596

There are nesting checks for when you're comparing a wet core to some other core. But for the type check at the call site see `++  fire`:

https://github.com/urbit/arvo/blob/master/sys/hoon.hoon#L9896

**Comment by philipcmonk** [comment]
Actually, if you added the "actual sample must nest in formal sample" check, then I think you could let wet gates nest within each other and even within dry gates in the same situations as dry gates, only with the type of the formal gate.  Not sure if dry gates would nest within them.

This would make them into a sort of bounded parametric polymorphism.  They'd still be pretty finicky, though, because there'd be no way extract the type parameters into something you could use.  So a function from `(list A)` to `(set A)` would have to avoid all internal typecasts or extract the type like wet gates do now.

**Comment by ohAitch** [comment]
Oh, I think this points towards why there isn't already nest checking: ?-
vain checking is disabled in wet gates, because sometimes you want to say
%foo for "%foo, or %bar, but definitely not any @tas" without having to
enumerate an explicit ?(%foo %bar %etc) type(or rather, using the wet gate
to create this enumeration). I suppose app/hood fails "code written not by
me", and is probably possible to recover in the proposed regime by adding
enough boilerplate; but it's a data point for why it won't "just" work.

**Comment by ohAitch** [comment]
And to reiterate, I don't think you want to enforce /cores/ nesting,
because then wet gates you pass to wet gates have to nest, and I don't
think you can tie that recursive knot even for something simple like
|*([a=gate b=noun] (a b)) without doing something weird = hard to follow.

(Briefly, without backwards inference / unification you'll have trouble
expressing that b doesn't need to be $-(* *) but rather only $-(_a *))
